[Merged by Bors] - feat(Data/Matroid/Loop): matroid loops

apnelson1 opened the pull request:
We add a definition of a loop element of a matroid, with basic API. 

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta commented on Mathlib/Data/Matroid/Loop.lean:
```diff
@@ -0,0 +1,176 @@
+# Terminology
+The set `M.closure ∅ = {e | M.IsLoop e}` appears frequently in statements about loops.
+we stick with the more convenient spelling `M.closure ∅`,
+but refer to it in lemma names by the shorter term `loops`.
```
Bearing this in mind, I wonder if it's any easier to define M.loops as this set, and use the spelling of being in that set for the predicate of being a loop. What do you think?

apnelson1 replied:
I considered this. (I think) the reason I didn't do this for my repo is so that the set of loops was more transparently something I could apply closure lemmas to. But it looks like I have helper lemmas like `loops_subset_closure` to abstract these uses out anyway. So this is probably a good idea for saving keystrokes as well as clarity. I'l make the change.

b-mehta commented on Mathlib/Data/Matroid/Loop.lean:
```diff
@@ -0,0 +1,176 @@
+/-!
+# Matroid Loops
+A 'loop' of a matroid `M` is an element `e` satisfying one of the following equivalent conditions:
+* `e ∈ M.closure ∅`;
+* `{e}` is dependent in `M`;
+* `{e}` is a circuit of `M`;
+* no base of `M` contains `e`.
```
It would be nice to state this using `TFAE`, and might save you some rewrites in practice

apnelson1 replied:
Done!

b-mehta reviewed (commented):
Can be a followup PR, but it'd be nice if the docs in Matroid.Restrict and Matroid.Constructions linked to the def here. 

I wonder if `loopyOn` should be defined in this file too.

I'll look at this again after you've made the change we discussed above, to save us both some potential duplicate work!

apnelson1 commented:
> Can be a followup PR, but it'd be nice if the docs in Matroid.Restrict and Matroid.Constructions linked to the def here.
> 
> I wonder if `loopyOn` should be defined in this file too.
> 
> I'll look at this again after you've made the change we discussed above, to save us both some potential duplicate work!

Thanks for the review

> Can be a followup PR, but it'd be nice if the docs in Matroid.Restrict and Matroid.Constructions linked to the def here.
> 
> I wonder if `loopyOn` should be defined in this file too.
> 
> I'll look at this again after you've made the change we discussed above, to save us both some potential duplicate work!



> Can be a followup PR, but it'd be nice if the docs in Matroid.Restrict and Matroid.Constructions linked to the def here.
> 
> I wonder if `loopyOn` should be defined in this file too.
> 
> I'll look at this again after you've made the change we discussed above, to save us both some potential duplicate work!

It would be a little annoying to move `loopyOn`, since it is used in the definition of `freeOn`, which then would cause quite a few things unrelated to loops to depend on the content of this file. I've added references to `IsLoop` in earlier docstrings as you suggest. 

(I also cleaned up a bunch of silly `Restriction` -> `IsRestriction`-related typos in the `Matroid.Restrict` module docstring, which I hope you can permit as a reviewer).

jcommelin reviewed (commented):
Thanks :tada:

bors merge
